snuba/stateful_consumer/states/bootstrap.py

```diff
@@ -1,25 +1,178 @@
+    """
+
+    def __init__(self):
+        # This represents the snapshot id we are currently tracking. It is set after
+        # snapshot-init and reset by snapshot-abort or after the current snapshot is
+        # loaded and a new init message is found.
+        self.__active_snapshot_id: Optional[SnapshotId] = None
+        # This tracks the transaction metadata for the currently loaded snapshot.
+        # it is set after a snapshot-load message.
+        self.__loaded_snapshot_transactions: Optional[TransactionData] = None
+        # Past processed snapshots. This is meant to tie abort and loaded messages
+        # to an init message and filter out snapshot messages we are not interested
+        # into.
+        self.__past_snapshots = set()
+
+    def get_completion_event(self) -> ConsumerStateCompletionEvent:
+        """
+        Returns the Completion State that drives the next state transition according
+        to the list of consumed messages.
+        """
+        if not self.__active_snapshot_id:
+            return ConsumerStateCompletionEvent.NO_SNAPSHOT
+        elif not self.__loaded_snapshot_transactions:
+            """
+            This means we received an init message but we are still waiting for
+            the snapshot loaded message.
+            """
+            return ConsumerStateCompletionEvent.SNAPSHOT_INIT_RECEIVED
+        else:
+            return ConsumerStateCompletionEvent.SNAPSHOT_READY_RECEIVED
+
+    def get_active_snapshot(self) -> Optional[Tuple[SnapshotId, Optional[TransactionData]]]:
+        if self.__active_snapshot_id:
+            return (self.__active_snapshot_id, self.__loaded_snapshot_transactions)
+        else:
+            return None
+
+    def __commit_if_no_active_snapshot(self) -> CommitDecision:
+        if not self.__active_snapshot_id:
+            return CommitDecision.COMMIT_THIS
+        else:
+            return CommitDecision.DO_NOT_COMMIT
+
+    def process_init(self, msg: SnapshotInit) -> CommitDecision:
+        """
+        Processes a snapshot-init message found on the control topic.
+        If the snapshot-init message is for this consumer and we are not already
+        waiting on an existing snapshot this sets the active snapshot and does not
+        commit the message for failover.
+        """
+        logger.debug("Processing init message for %r", msg.id)
+        if msg.product != settings.SNAPSHOT_LOAD_PRODUCT:
+            return self.__commit_if_no_active_snapshot()
+        if msg.id in self.__past_snapshots:
+            logger.warning(
+                "Duplicate Snapshot init: %r",
+                msg.id,
+            )
+            return self.__commit_if_no_active_snapshot()
+        if self.__active_snapshot_id and not self.__loaded_snapshot_transactions:
+            logger.error(
+                "A new snapshot-init message was found before a previous one "
+                " was completed. Ignoring. Running %r. Init received %r.",
```

Comment:
"running" here still strikes me as odd, since it doesn't change the active snapshot. Would it make more sense to word this as `Ignoring %r, and continuing with %r` instead?

It also seems a bit weird to use `error` when this isn't logging an exception, kind of feels like this should be a `warning` instead.